clean_code/algorithms/creation_algorithms.py

```diff
@@ -0,0 +1,68 @@
+from math import factorial
+
+"""This module supplies fast algorithms for pattern-related creations:
+   all permutations of a staring, duplicate-deletion."""
+
+def create_trimmed_string(base_string):
+    """deletes duplicates in a string. example: aaabbaac -> abac"""
+    trimmed_elements = [base_string[0]]
+    for index in range(1, len(base_string)):
+        if base_string[index] != base_string[index-1]:
+            trimmed_elements.append(base_string[index])
+    return ''.join(trimmed_elements)
+
+
+def generate_permutations_iterably(elements):
+    """returns all the permutations of a string. the function is non-recursive"""
+    permutations = set()
+    final_permutations = set()
+    next_index_to_be_permutated = 0
+    expectd_permutations_number = factorial(len(elements))
+
+    new_permutations = _generate_new_permutations(elements)
+    permutations.update(new_permutations)
+
+    while len(permutations) < expectd_permutations_number:
+        permutations_copy = permutations.copy()
+        permutations = set()
+
+        for permutation in permutations_copy:
+            new_permutations = _generate_new_permutations(permutation)
+            final_permutations.update(new_permutations)
+            permutations.update(new_permutations)
+
+    return [''.join(permutation) for permutation in final_permutations]
```

Comment:
You are doing this conversion twice: here and at `_generate_new_permutations`. Why? maybe both places is not the true and another function is required?